dlt/common/libs/pyarrow.py

```diff
@@ -654,104 +742,210 @@ def row_tuples_to_arrow(
-                " extracting an SQL VIEW that selects with cast."
-            )
-            json_str_array = pa.array(
-                [
-                    (
-                        None
-                        if s is None
-                        else json.dumps(s) if not issubclass(type(s), set) else json.dumps(list(s))
-                    )
-                    for s in columnar_known_types[field.name]
-                ]
+                "Type can't be inferred by `pyarrow`. Values will be encoded as in a loop, slowing"
+                " extraction."
             )
-            columnar_known_types[field.name] = json_str_array
-
-    # If there are unknown type columns, first create a table to infer their types
-    if columnar_unknown_types:
-        new_schema_fields = []
-        for key in list(columnar_unknown_types):
-            arrow_col: Optional[pa.Array] = None
-            try:
-                arrow_col = pa.array(columnar_unknown_types[key])
-                if pa.types.is_null(arrow_col.type):
-                    logger.warning(
-                        f"Column {key} contains only NULL values and data type could not be"
-                        " inferred. This column is removed from a arrow table"
-                    )
+            encoded_values: list[Union[None, Mapping[Any, Any], Sequence[Any], str]] = []
+            for value in column_data:
+                if value is None:
+                    encoded_values.append(None)
                     continue
-
-            except pa.ArrowInvalid as e:
-                # Try coercing types not supported by arrow to a json friendly format
-                # E.g. dataclasses -> dict, UUID -> str
                 try:
-                    arrow_col = pa.array(
-                        map_nested_in_place(custom_encode, list(columnar_unknown_types[key]))
-                    )
-                    logger.warning(
-                        f"Column {key} contains a data type which is not supported by pyarrow and"
-                        f" got converted into {arrow_col.type}. This slows down arrow table"
-                        " generation."
-                    )
-                except (pa.ArrowInvalid, TypeError):
-                    logger.warning(
-                        f"Column {key} contains a data type which is not supported by pyarrow. This"
-                        f" column will be ignored. Error: {e}"
-                    )
-            if arrow_col is not None:
-                columnar_known_types[key] = arrow_col
-                new_schema_fields.append(
-                    pa.field(
-                        key,
-                        arrow_col.type,
-                        nullable=columns[key].get("nullable", True),
-                    )
-                )
+                    # the 3 types match those supported by `map_nested_in_place()`
+                    if isinstance(value, (tuple, dict, list)):
```

Comment:
if we see dict or list we should do what we do in case (3) - json.dumpb. for tuples and sets - convert to list.

other values: as below (custom_encode but then encode to utf-8 so we have consistently bytes).

tldr;> dump nested types to json. or maybe I miss something here?